### PR TITLE
[EP-559] Default card functionality

### PR DESCRIPTION
**Pull request by gcojocaru**
- added default card functionality.

### CODE REVIEW

File: Source/Models/CardStorage/JPCardStorage.h [hunk 1]
```diff
@@ -89,4 +89,10 @@
  */
 - (void)setCardAsSelectedAtIndex:(NSUInteger)index;
 
+/**
+* A method that sets a card as default based on an index
+*/
+- (void)setCardAsDefaultAtIndex:(NSInteger)index;
```
Comment (mpetrenco) [inline Source/Models/CardStorage/JPCardStorage.h]: Please use `NSUInteger` instead

File: Source/Models/CardStorage/JPCardStorage.h [hunk 2]
```diff
@@ -89,4 +89,10 @@
...
+
+
```
Comment (mpetrenco) [inline Source/Models/CardStorage/JPCardStorage.h]: Please run `./clang-format.sh`

File: Source/Models/CardStorage/JPCardStorage.m [hunk 3]
```diff
@@ -111,4 +111,30 @@ - (void)setCardAsSelectedAtIndex:(NSUInteger)index {
...
+    JPStoredCardDetails *defaultCard;
+    for (JPStoredCardDetails *card in self.storedCards) {
+        if(card.isDefault) {
+            defaultCard = card;
+            break;
+        }
+        defaultCardIndex++;
+    }
+    if (defaultCardIndex>=0 && defaultCard) {
+        [self.storedCards removeObjectAtIndex:defaultCardIndex];
```
Comment (mpetrenco) [inline Source/Models/CardStorage/JPCardStorage.m]: There should be an `update card at index` method that you can use instead

File: Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.h [hunk 4]
```diff
@@ -50,6 +50,13 @@
  */
 - (void)updateStoredCardTitleWithInput:(nonnull NSString *)input;
 
+/**
+ * A method that updates the stored card's isDefault property based on a provided input string
+ *
+ * @param isDefualt - a boolean value that indicates if the card is selected as default card
+ */
+-(void)updateStoredCardDefaultWithValue:(BOOL)isDefualt;
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.h]: `isDefualt`

File: Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.h [hunk 5]
```diff
@@ -50,6 +50,13 @@
  */
 - (void)updateStoredCardTitleWithInput:(nonnull NSString *)input;
 
+/**
+ * A method that updates the stored card's isDefault property based on a provided input string
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.h]: `based on a provided input string` ?

File: Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.m [hunk 6]
```diff
@@ -60,4 +60,10 @@ - (void)updateStoredCardTitleWithInput:(NSString *)input {
                                             atIndex:self.cardIndex];
 }
 
+-(void)updateStoredCardDefaultWithValue:(BOOL)isDefualt{
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.m]: `isDefualt`

Comment (mpetrenco) [reply Source/Modules/CardCustomization/Interactor/JPCardCustomizationInteractor.m]: Why are we passing a `isDefault` parameter if it doesn't do anything unless it is `YES` ?
Can we use ` - (void)updateDefaultStoredCard;` instead?

File: Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h [hunk 7]
```diff
@@ -65,6 +65,11 @@
  * This method saves the applied changes in the keychain and dismisses the Card Customization screen.
  */
 - (void)handleSaveEvent;
+/**
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h]: Add a space here

File: Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h [hunk 8]
```diff
@@ -65,6 +65,11 @@
...
+* A method that is triggered once the once the cell with that sets default cart is tapped by a user
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h]: `once the once the`

Comment (mpetrenco) [reply Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h]: `once the cell with that sets default cart` ?

File: Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h [hunk 9]
```diff
@@ -65,6 +65,11 @@
...
+* This method changes selection state in viewModel and reloads the view with the new values.
+*/
+-(void)handleRadioButtonEvent;
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.h]: This is a method that triggers specifically when the user taps on the `isDefault` cell.
`handleRadioButtonEvent` is too generic. Please rename to `handleToggleDefaultCardEvent` or something similar

File: Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.m [hunk 10]
```diff
@@ -40,6 +40,7 @@ @interface JPCardCustomizationPresenterImpl ()
 @property (nonatomic, strong) JPCardCustomizationTextInputModel *textInputModel;
 @property (nonatomic, strong) JPCardCustomizationIsDefaultModel *isDefaultModel;
 @property (nonatomic, strong) JPCardCustomizationSubmitModel *submitModel;
+@property (nonatomic, assign) BOOL isSelectredCardDefault;
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.m]: `isSelectred`
Also can we rename to `isDefaultCardSelected`

File: Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.m [hunk 11]
```diff
@@ -84,9 +85,19 @@ - (void)handleCancelEvent {
 - (void)handleSaveEvent {
     [self.interactor updateStoredCardTitleWithInput:self.selectedCardTitle];
     [self.interactor updateStoredCardPatternWithType:self.selectedPatternType];
+    [self.interactor updateStoredCardDefaultWithValue:self.isDefaultModel.isDefault];
     [self.router navigateBack];
 }
 
+-(void)handleRadioButtonEvent{
+    self.isSelectredCardDefault = !self.isDefaultModel.isDefault;
```
Comment (mpetrenco) [inline Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.m]: Why do we need the `isSelectredCardDefault ` variable?
Can't we just use the `isDefaultModel.isDefault` toggle instead?

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.h [hunk 12]
```diff
@@ -75,6 +75,13 @@
  */
 - (void)setCardAsSelectedAtInded:(NSUInteger)index;
 
+
+/**
+* A method that sets a card as default based on an index
+*/
+- (void)setCardAsDefaultAtIndex:(NSInteger)index;
+
+
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.h]: Extra space here

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 13]
```diff
@@ -85,6 +85,13 @@ - (void)selectCardAtIndex:(NSUInteger)index {
     }
 }
 
+
+#pragma mark - Set cart at index as default
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m]: `cart`

File: Source/Modules/PaymentMethods/Presenter/JPPaymentMethodsPresenter.m [hunk 14]
```diff
@@ -110,6 +111,35 @@ - (void)handleApplePayButtonTap {
     }];
 }
 
+- (void)handlePaymentResponse:(JPResponse *)response {
+    NSInteger selectedCardIndex = [self indexOfSelectdCard];
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Presenter/JPPaymentMethodsPresenter.m]: `selectd`

Comment (mpetrenco) [reply Source/Modules/PaymentMethods/Presenter/JPPaymentMethodsPresenter.m]: If no card is selected, then the `[self indexOfSelectCard]` returns `-1`.
You pass this to the interactor without handling this case, crashing the app.

File: Source/Services/CardValidation/JPCardValidationService.m [hunk 15]
```diff
@@ -29,6 +29,7 @@
 #import "JPValidationResult.h"
 #import "NSError+Additions.h"
 #import "NSString+Additions.h"
+#import "JPConstants.h"
```
Comment (mpetrenco) [inline Source/Services/CardValidation/JPCardValidationService.m]: Do you need `JPConstants` here?

File: Tests/Modules/CardStorage/CardStorageTest.swift [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        let thirdStoredCard = JPStoredCardDetails(lastFour: "3333", expiryDate: "23/23", cardNetwork: .networkJCB, cardToken: "cardToken3")
+        let forthStoredCard = JPStoredCardDetails(lastFour: "4444", expiryDate: "24/24", cardNetwork: .networkAMEX, cardToken: "cardToken4")
+        
+        JPCardStorage.sharedInstance()?.deleteCardDetails()
+        JPCardStorage.sharedInstance()!.add(firstStoredCard)
+        JPCardStorage.sharedInstance()!.add(secondStoredCard)
+        JPCardStorage.sharedInstance()!.add(thirdStoredCard)
+        JPCardStorage.sharedInstance()!.add(forthStoredCard)
+
+
```
Comment (mpetrenco) [inline Tests/Modules/CardStorage/CardStorageTest.swift]: Extra spaces

File: Source/Services/CardValidation/JPCardValidationService.m (hunk 15 above)
Comment (gcojocaru) [reply Source/Services/CardValidation/JPCardValidationService.m]: YES

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.h [hunk 17]
```diff
@@ -90,6 +94,11 @@
 - (void)handle3DSecureTransactionFromError:(NSError *)error
                                 completion:(JudoCompletionBlock)completion;
 
+/**
+ * A method that reorders cards so that the default card is always on top
+*/
+- (void)orederCards;
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.h]: `oreder`

File: Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.m [hunk 18]
```diff
@@ -84,9 +82,18 @@ - (void)handleCancelEvent {
 - (void)handleSaveEvent {
     [self.interactor updateStoredCardTitleWithInput:self.selectedCardTitle];
     [self.interactor updateStoredCardPatternWithType:self.selectedPatternType];
+    [self.interactor updateStoredCardDefaultWithValue:self.isDefaultModel.isDefault];
     [self.router navigateBack];
 }
 
+- (void)handleToggleDefaultCardEvent {
+    self.isDefaultModel.isDefault = !self.isDefaultModel.isDefault;
```
Comment (stpopa) [inline Source/Modules/CardCustomization/Presenter/JPCardCustomizationPresenter.m]: can we get rid of the repetition `isDefault` here? 
`self.isDefaultModel.isDefault`

**Pull request merged**